src/k8s/pkg/k8sd/controllers/node_configuration.go

```diff
@@ -0,0 +1,82 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/canonical/k8s/pkg/k8sd/types"
+	"github.com/canonical/k8s/pkg/snap"
+	snaputil "github.com/canonical/k8s/pkg/snap/util"
+	"github.com/canonical/k8s/pkg/utils/k8s"
+	"github.com/mitchellh/mapstructure"
+	v1 "k8s.io/api/core/v1"
+)
+
+type NodeConfigurationController struct {
+}
+
+func NewNodeConfigurationController() *NodeConfigurationController {
+	return &NodeConfigurationController{}
+}
+
+func (c *NodeConfigurationController) Run(ctx context.Context, snap snap.Snap) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
```

Comment:
change the `time.Sleep()` at the end of the loop with a `<-time.After(3 * time.Second)` here. This interval should be something larger like 10 seconds (or more?). That way you don't need multiple `time.Sleep()`s in the loop, and you don't have to wait for sleep to finish to see that the context has been canceled.

Reply:
I'm not exactly sure if the interval should be something larger. Mainly 2 concerns:
* The initial start of the controller might be a bit delayed, which increases the time for nodes to settle.
* If for some reason the `Watch` fails, it will take 10 seconds to recover

I think a small interval is okay, the only worry I would have is for the failure on getting a kubernetes client when the cluster is not bootstrapped/joined. Especially in the scenario where someone installed the `k8s` snap and never interacted with it. We would print an error message into the logs every 3 seconds, that can eventually exhaust disk space.